src/EditorFeatures/Core/BraceMatching/AbstractEmbeddedLanguageBraceMatcher.cs

```diff
@@ -2,44 +2,154 @@
+                SemanticModel semanticModel,
+                int position,
+                BraceMatchingOptions options,
+                ArrayBuilder<IEmbeddedLanguageBraceMatcher> serviceBuffer,
+                CancellationToken cancellationToken)
             {
-                foreach (var language in languagesProvider.Languages)
+                _service = service;
+                _semanticModel = semanticModel;
+                _position = position;
+                _options = options;
+                _serviceBuffer = serviceBuffer;
+                _cancellationToken = cancellationToken;
+            }
+
+            public BraceMatchingResult? Recurse(SyntaxNode node)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+                if (node.Span.IntersectsWith(_position))
                 {
-                    var braceMatcher = (language as IEmbeddedLanguageEditorFeatures)?.BraceMatcher;
-                    if (braceMatcher != null)
+                    foreach (var child in node.ChildNodesAndTokens())
                     {
-                        var result = await braceMatcher.FindBracesAsync(
-                            document, position, options, cancellationToken).ConfigureAwait(false);
-                        if (result != null)
+                        if (child.IsNode)
                         {
-                            return result;
+                            var value = Recurse(child.AsNode()!);
+                            if (value.HasValue)
+                                return value;
+                        }
+                        else
+                        {
+                            var value = ProcessToken(child.AsToken());
+                            if (value.HasValue)
+                                return value;
                         }
                     }
                 }
+
+                return null;
             }
 
-            return null;
+            private BraceMatchingResult? ProcessToken(SyntaxToken token)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+                return ClassifyToken(token);
+            }
+
+            private BraceMatchingResult? ClassifyToken(SyntaxToken token)
+            {
+                if (token.Span.IntersectsWith(_position) && _service.SyntaxTokenKinds.Contains(token.RawKind))
+                {
+                    _serviceBuffer.Clear();
+
+                    // First, see if this is a string annotated with either a comment or [StringSyntax] attribute. If
+                    // so, delegate to the first classifier we have registered for whatever language ID we find.
+                    if (_service.Detector.IsEmbeddedLanguageToken(token, _semanticModel, _cancellationToken, out var identifier, out _) &&
+                        _service.IdentifierToServices.TryGetValue(identifier, out var services))
```

Comment:
names are confusing here - we have a `_service` which we're getting `services` from?  Then iterate through `service`

Reply:
Done.